Liquibase started to throw exception after upgrading to version 4.12.0

The [POC Project](https://github.com/liquibase/liquibase/files/8948917/POC.zip) has been added as attachment which has a similar approach with the pre-existing project.

Previously we were using liquibase-core version 4.5.0. We decided to upgrade it to 4.12.0 because of one of the [known vulnerability](https://nvd.nist.gov/vuln/detail/CVE-2022-0839) which has resolved starting from version 4.8.0.

* From the root path of the project I execute the following command to start liquibase migration
* `mvn compile package && java -jar target/test-2.6.6.jar`
* I've got the following error :

<Details><Summary>Please extend this to see the error log</Summary>

```
[INFO] Scanning for projects...
[INFO]
[INFO] --------------------------< com.example:test >--------------------------
[INFO] Building Liquibase Bug 2.6.6
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-resources-plugin:3.2.0:resources (default-resources) @ test ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] Copying 1 resource
[INFO] Copying 3 resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ test ---
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-resources-plugin:3.2.0:resources (default-resources) @ test ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] Copying 1 resource
[INFO] Copying 3 resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ test ---
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-resources-plugin:3.2.0:testResources (default-testResources) @ test ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] skip non existing resourceDirectory LiquibaseBug/src/test/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ test ---
[INFO] No sources to compile
[INFO]
[INFO] --- maven-surefire-plugin:2.22.2:test (default-test) @ test ---
[INFO] No tests to run.
[INFO]
[INFO] --- maven-jar-plugin:3.2.2:jar (default-jar) @ test ---
[INFO] Building jar: LiquibaseBug/target/test-2.6.6.jar
[INFO]
[INFO] --- spring-boot-maven-plugin:2.6.6:repackage (repackage) @ test ---
[INFO] Replacing main artifact with repackaged archive
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.641 s
[INFO] Finished at: 2022-06-21T15:36:34+03:00
[INFO] ------------------------------------------------------------------------

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::                (v2.6.6)

2022-06-21 15:36:35.654  INFO 23791 --- [           main] com.example.BootAppConfig                : Starting BootAppConfig v2.6.6 using Java 11.0.8
2022-06-21 15:36:35.658  INFO 23791 --- [           main] com.example.BootAppConfig                : No active profile set, falling back to 1 default profile: "default"
2022-06-21 15:36:36.318  INFO 23791 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data JPA repositories in DEFAULT mode.
2022-06-21 15:36:36.336  INFO 23791 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 6 ms. Found 0 JPA repository interfaces.
2022-06-21 15:36:36.845  INFO 23791 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Starting...
2022-06-21 15:36:37.565  INFO 23791 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Start completed.
2022-06-21 15:36:37.926  INFO 23791 --- [           main] liquibase.database                       : Set default schema name to public
2022-06-21 15:36:39.045  INFO 23791 --- [           main] liquibase.lockservice                    : Successfully acquired change log lock
2022-06-21 15:36:40.107  INFO 23791 --- [           main] liquibase.changelog                      : Reading from public.databasechangelog
Running Changeset: db/changelog/migrations/db.bug.xml::12345678-2::test
2022-06-21 15:36:41.784  INFO 23791 --- [           main] liquibase.lockservice                    : Successfully released change log lock
2022-06-21 15:36:41.942  WARN 23791 --- [           main] s.c.a.AnnotationConfigApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'liquibase' defined in class path resource [org/springframework/boot/autoconfigure/liquibase/LiquibaseAutoConfiguration$LiquibaseConfiguration.class]: Invocation of init method failed; nested exception is liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for change set db/changelog/migrations/db.bug.xml::12345678-2::test:
     Reason: liquibase.exception.DatabaseException: Unterminated string literal started at position 43 in SQL INSERT INTO script (id, script) VALUES (7, '{test. Expected  char [Failed SQL: (0) INSERT INTO script (id, script) VALUES (7, '{test]
2022-06-21 15:36:41.942  INFO 23791 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Shutdown initiated...
2022-06-21 15:36:41.951  INFO 23791 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Shutdown completed.
2022-06-21 15:36:41.963  INFO 23791 --- [           main] ConditionEvaluationReportLoggingListener :

Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2022-06-21 15:36:41.991 ERROR 23791 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'liquibase' defined in class path resource [org/springframework/boot/autoconfigure/liquibase/LiquibaseAutoConfiguration$LiquibaseConfiguration.class]: Invocation of init method failed; nested exception is liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for change set db/changelog/migrations/db.bug.xml::12345678-2::test:
     Reason: liquibase.exception.DatabaseException: Unterminated string literal started at position 43 in SQL INSERT INTO script (id, script) VALUES (7, '{test. Expected  char [Failed SQL: (0) INSERT INTO script (id, script) VALUES (7, '{test]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1804) ~[spring-beans-5.3.18.jar!/:5.3.18]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:620) ~[spring-beans-5.3.18.jar!/:5.3.18]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542) ~[spring-beans-5.3.18.jar!/:5.3.18]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335) ~[spring-beans-5.3.18.jar!/:5.3.18]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-5.3.18.jar!/:5.3.18]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333) ~[spring-beans-5.3.18.jar!/:5.3.18]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208) ~[spring-beans-5.3.18.jar!/:5.3.18]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:322) ~[spring-beans-5.3.18.jar!/:5.3.18]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208) ~[spring-beans-5.3.18.jar!/:5.3.18]
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1154) ~[spring-context-5.3.18.jar!/:5.3.18]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:908) ~[spring-context-5.3.18.jar!/:5.3.18]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583) ~[spring-context-5.3.18.jar!/:5.3.18]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:740) ~[spring-boot-2.6.6.jar!/:2.6.6]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:415) ~[spring-boot-2.6.6.jar!/:2.6.6]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:303) ~[spring-boot-2.6.6.jar!/:2.6.6]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1312) ~[spring-boot-2.6.6.jar!/:2.6.6]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1301) ~[spring-boot-2.6.6.jar!/:2.6.6]
	at com.example.BootAppConfig.main(BootAppConfig.java:24) ~[classes!/:2.6.6]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49) ~[test-2.6.6.jar:2.6.6]
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:108) ~[test-2.6.6.jar:2.6.6]
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:58) ~[test-2.6.6.jar:2.6.6]
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:88) ~[test-2.6.6.jar:2.6.6]
Caused by: liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for change set db/changelog/migrations/db.bug.xml::12345678-2::test:
     Reason: liquibase.exception.DatabaseException: Unterminated string literal started at position 43 in SQL INSERT INTO script (id, script) VALUES (7, '{test. Expected  char [Failed SQL: (0) INSERT INTO script (id, script) VALUES (7, '{test]
	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:126) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Liquibase.lambda$null$0(Liquibase.java:265) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.lambda$child$0(Scope.java:180) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:189) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:179) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:158) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:243) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Liquibase.lambda$update$1(Liquibase.java:264) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.lambda$child$0(Scope.java:180) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:189) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:179) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:158) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Liquibase.runInScope(Liquibase.java:2405) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Liquibase.update(Liquibase.java:211) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Liquibase.update(Liquibase.java:197) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.integration.spring.SpringLiquibase.performUpdate(SpringLiquibase.java:314) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.integration.spring.SpringLiquibase.afterPropertiesSet(SpringLiquibase.java:269) ~[liquibase-core-4.12.0.jar!/:na]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1863) ~[spring-beans-5.3.18.jar!/:5.3.18]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1800) ~[spring-beans-5.3.18.jar!/:5.3.18]
	... 25 common frames omitted
Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set db/changelog/migrations/db.bug.xml::12345678-2::test:
     Reason: liquibase.exception.DatabaseException: Unterminated string literal started at position 43 in SQL INSERT INTO script (id, script) VALUES (7, '{test. Expected  char [Failed SQL: (0) INSERT INTO script (id, script) VALUES (7, '{test]
	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:696) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:56) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.changelog.ChangeLogIterator$2.lambda$null$0(ChangeLogIterator.java:113) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.lambda$child$0(Scope.java:180) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:189) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:179) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:158) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.changelog.ChangeLogIterator$2.lambda$run$1(ChangeLogIterator.java:112) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.lambda$child$0(Scope.java:180) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:189) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:179) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:158) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:243) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.changelog.ChangeLogIterator$2.run(ChangeLogIterator.java:93) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.lambda$child$0(Scope.java:180) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:189) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:179) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:158) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:243) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.Scope.child(Scope.java:247) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:65) ~[liquibase-core-4.12.0.jar!/:na]
	... 43 common frames omitted
Caused by: liquibase.exception.DatabaseException: Unterminated string literal started at position 43 in SQL INSERT INTO script (id, script) VALUES (7, '{test. Expected  char [Failed SQL: (0) INSERT INTO script (id, script) VALUES (7, '{test]
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:434) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:77) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:160) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.database.AbstractJdbcDatabase.execute(AbstractJdbcDatabase.java:1296) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.database.AbstractJdbcDatabase.executeStatements(AbstractJdbcDatabase.java:1278) ~[liquibase-core-4.12.0.jar!/:na]
	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:661) ~[liquibase-core-4.12.0.jar!/:na]
	... 63 common frames omitted
Caused by: org.postgresql.util.PSQLException: Unterminated string literal started at position 43 in SQL INSERT INTO script (id, script) VALUES (7, '{test. Expected  char
	at org.postgresql.core.Parser.checkParsePosition(Parser.java:1380) ~[postgresql-42.3.3.jar!/:42.3.3]
	at org.postgresql.core.Parser.parseSql(Parser.java:1287) ~[postgresql-42.3.3.jar!/:42.3.3]
	at org.postgresql.core.Parser.replaceProcessing(Parser.java:1231) ~[postgresql-42.3.3.jar!/:42.3.3]
	at org.postgresql.core.CachedQueryCreateAction.create(CachedQueryCreateAction.java:43) ~[postgresql-42.3.3.jar!/:42.3.3]
	at org.postgresql.core.QueryExecutorBase.createQueryByKey(QueryExecutorBase.java:340) ~[postgresql-42.3.3.jar!/:42.3.3]
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:307) ~[postgresql-42.3.3.jar!/:42.3.3]
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:291) ~[postgresql-42.3.3.jar!/:42.3.3]
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:286) ~[postgresql-42.3.3.jar!/:42.3.3]
	at com.zaxxer.hikari.pool.ProxyStatement.execute(ProxyStatement.java:94) ~[HikariCP-4.0.3.jar!/:na]
	at com.zaxxer.hikari.pool.HikariProxyStatement.execute(HikariProxyStatement.java) ~[HikariCP-4.0.3.jar!/:na]
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:430) ~[liquibase-core-4.12.0.jar!/:na]
	... 68 common frames omitted
```

</Details>

I am using the following changesets:

<Details><Summary>Please extend this to see changesets</Summary>

```

<?xml version="1.0" encoding="UTF-8"?>

<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.8.xsd">
    <changeSet author="test" id="123456789">
        <createTable tableName="script">
            <column name="id" type="BIGINT">
                <constraints nullable="false" primaryKey="true" primaryKeyName="script_pkey"/>
            </column>
            <column name="script" type="TEXT"/>
        </createTable>
    </changeSet>

    <!-- Doesn't work with liquibase-core version 4.12-->
    <!-- Works with liquibase-core version 4.11-->
    <changeSet id="12345678-2" author="test">
        <sql>
            INSERT INTO script (id, script) VALUES (4, '${test}');
            INSERT INTO script (id, script) VALUES (7, '{test}');
        </sql>
    </changeSet>

    <!-- Doesn't work with liquibase-core version 4.12-->
    <!-- Works with liquibase-core version 4.11-->
    <changeSet id="12345678-3" author="test">
        <sql>
            INSERT INTO script (id, script) VALUES (1, '${test} {test}');
        </sql>
    </changeSet>

    <!-- Works with liquibase-core version 4.12-->
    <!-- Works with liquibase-core version 4.11-->
    <changeSet id="12345678-4" author="test">
        <sql>
            INSERT INTO script (id, script) VALUES (5, '{test} ${test}');
        </sql>
    </changeSet>

    <!-- Works with liquibase-core version 4.12-->
    <!-- Works with liquibase-core version 4.11-->
    <changeSet id="12345678-5" author="test">
        <sql>
            INSERT INTO script (id, script)  VALUES (8, '{test}');
        </sql>
    </changeSet>

    <!-- Works with liquibase-core version 4.12-->
    <!-- Works with liquibase-core version 4.11-->
    <changeSet id="12345678-6" author="test">
        <sql>
            INSERT INTO script (id, script) VALUES (10, '${test}');
        </sql>
    </changeSet>

</databaseChangeLog>


```

</Details>

--- a/liquibase-core/src/main/java/liquibase/changelog/ExpressionExpander.java
+++ b/liquibase-core/src/main/java/liquibase/changelog/ExpressionExpander.java
@@ -30,18 +30,14 @@ class ExpressionExpander {
             int nextChar = reader.read();
             while (nextChar != -1) {
                 if (nextChar == '$') {
-                    reader.mark(1);
                     if (reader.read() == '{') {
-                        reader.mark(1);
-                        if (enableEscaping && reader.read() == ':') {
-                            stringBuilder.append("${");
-                            stringBuilder.append(expandExpressions(reader, changeLog, true).trim());
-                            stringBuilder.append("}");
+                        String paramExpression = expandExpressions(reader, changeLog, true);
+
+                        Object paramValue;
+                        if (paramExpression.startsWith(":") && enableEscaping) {
+                            paramValue = "${" + paramExpression.substring(1).trim() + "}";
                         } else {
-                            inExpression = true;
-                            reader.reset();
-                            String paramExpression = expandExpressions(reader, changeLog, true);
-                            Object paramValue = parameters.getValue(paramExpression.trim(), changeLog);
+                            paramValue = parameters.getValue(paramExpression.trim(), changeLog);
 
                             if (paramValue == null) {
                                 final ChangeLogParserConfiguration.MissingPropertyMode missingPropertyMode = ChangeLogParserConfiguration.MISSING_PROPERTY_MODE.getCurrentValue();
@@ -62,11 +58,13 @@ class ExpressionExpander {
                                     paramValue = expandExpressions((String) paramValue, changeLog);
                                 }
                             }
-                            stringBuilder.append(paramValue);
                         }
+
+                        stringBuilder.append(paramValue);
                     } else {
-                        reader.reset();
-                        stringBuilder.append("$");
+                        stringBuilder.append("${");
+                        nextChar = reader.read();
+                        continue;
                     }
                 } else {
                     if (nextChar == '}' && inExpression) {

--- a/liquibase-core/src/main/java/liquibase/changelog/ExpressionExpander.java
+++ b/liquibase-core/src/main/java/liquibase/changelog/ExpressionExpander.java
@@ -30,11 +30,14 @@ class ExpressionExpander {
             int nextChar = reader.read();
             while (nextChar != -1) {
                 if (nextChar == '$') {
+                    reader.mark(1);
                     if (reader.read() == '{') {
                         String paramExpression = expandExpressions(reader, changeLog, true);
 
                         Object paramValue;
-                        if (paramExpression.startsWith(":") && enableEscaping) {
+                        if (paramExpression.startsWith("${")) {
+                            paramValue = paramExpression; //was not actually a valid expression
+                        } else if (paramExpression.startsWith(":") && enableEscaping) {
                             paramValue = "${" + paramExpression.substring(1).trim() + "}";
                         } else {
                             paramValue = parameters.getValue(paramExpression.trim(), changeLog);
@@ -62,7 +65,8 @@ class ExpressionExpander {
 
                         stringBuilder.append(paramValue);
                     } else {
-                        stringBuilder.append("${");
+                        stringBuilder.append("$");
+                        reader.reset();
                         nextChar = reader.read();
                         continue;
                     }
@@ -80,6 +84,11 @@ class ExpressionExpander {
             // Unreachable as we initialize the StringReader with a non-null string
         }
 
-        return stringBuilder.toString();
+        if (inExpression) {
+            //never got to the trailing `}`, return the string as-is
+            return "${"+stringBuilder;
+        } else {
+            return stringBuilder.toString();
+        }
     }
 }

--- a/liquibase-core/src/test/groovy/liquibase/changelog/ExpressionExpanderTest.groovy
+++ b/liquibase-core/src/test/groovy/liquibase/changelog/ExpressionExpanderTest.groovy
@@ -69,6 +69,7 @@ class ExpressionExpanderTest extends Specification {
         PRESERVE | "valid '\${double.nested.value.valid}' double nested"                                             | "valid 'double a param with param1=value1' double nested"
         PRESERVE | "invalid '\${double.nested.value.invalid}' double nested"                                         | "invalid 'double \${nested.value.invalid}' double nested"
         EMPTY    | "invalid '\${double.nested.value.invalid}' double nested"                                         | "invalid 'double ' double nested"
+        PRESERVE | "INSERT INTO script (id, script) VALUES (5, '{test} \${test} {test} \${test} {test} ');"                                  | "INSERT INTO script (id, script) VALUES (5, '{test} \${test} {test} \${test} {test} ');"
     }
 
 

--- a/liquibase-core/src/test/groovy/liquibase/changelog/ExpressionExpanderTest.groovy
+++ b/liquibase-core/src/test/groovy/liquibase/changelog/ExpressionExpanderTest.groovy
@@ -44,6 +44,7 @@ class ExpressionExpanderTest extends Specification {
         PRESERVE | ""                                                                                                | ""
         PRESERVE | "A Simple String"                                                                                 | "A Simple String"
         PRESERVE | "With { brackets } in it"                                                                         | "With { brackets } in it"
+        PRESERVE | "With \${ unmatched"                                                                              | "With \${ unmatched"
         PRESERVE | "A string with one expression \${param1} set"                                                     | "A string with one expression value1 set"
         PRESERVE | "A string with two expressions \${param1} and \${param2} set"                                     | "A string with two expressions value1 and value2 set"
         EMPTY    | "A string with two expressions \${param1} and \${param2} set"                                     | "A string with two expressions value1 and value2 set"
@@ -69,7 +70,12 @@ class ExpressionExpanderTest extends Specification {
         PRESERVE | "valid '\${double.nested.value.valid}' double nested"                                             | "valid 'double a param with param1=value1' double nested"
         PRESERVE | "invalid '\${double.nested.value.invalid}' double nested"                                         | "invalid 'double \${nested.value.invalid}' double nested"
         EMPTY    | "invalid '\${double.nested.value.invalid}' double nested"                                         | "invalid 'double ' double nested"
-        PRESERVE | "INSERT INTO script (id, script) VALUES (5, '{test} \${test} {test} \${test} {test} ');"                                  | "INSERT INTO script (id, script) VALUES (5, '{test} \${test} {test} \${test} {test} ');"
+        PRESERVE | "INSERT INTO script (id, script) VALUES (5, '{test} \${test} {test} \${test} {test} ');"          | "INSERT INTO script (id, script) VALUES (5, '{test} \${test} {test} \${test} {test} ');"
+        PRESERVE | "test \$\$ here"                                                                                  | "test \$\$ here"
+        PRESERVE | "test \$1.30 here"                                                                                | "test \$1.30 here"
+        PRESERVE | "test \$1.30 here \$"                                                                             | "test \$1.30 here \$"
+        PRESERVE | "ending \${"                                                                                      | "ending \${"
+        PRESERVE | "ending \$"                                                                                       | "ending \$"
     }
 
 

